query-algebrizer/src/cc.rs

```diff
@@ -0,0 +1,795 @@
+                // from the type of the value itself! #292.
+                self.constrain_column_to_constant(alias.1.clone(), DatomsColumn::Value, c.clone().into_typed_value());
+            },
+        }
+
+    }
+
+    pub fn apply_pattern<'s, 'p>(&mut self, schema: &'s Schema, pattern: &'p Pattern) {
+        // For now we only support the default source.
+        match pattern.source {
+            Some(SrcVar::DefaultSrc) | None => (),
+            _ => unimplemented!(),
+        };
+
+        if let Some(alias) = self.alias_table(schema, &pattern) {
+            self.apply_pattern_clause_for_alias(schema, pattern, &alias);
+            self.from.push(alias);
+        } else {
+            // We didn't determine a table, likely because there was a mismatch
+            // between an attribute and a value.
+            // We know we cannot return a result, so we short-circuit here.
+            self.mark_known_empty("Table aliaser couldn't determine a table.");
+        }
+    }
+}
+
+#[cfg(test)]
+mod testing {
+    use super::*;
+
+    fn associate_ident(schema: &mut Schema, i: &str, e: Entid) {
+        schema.entid_map.insert(e, i.to_string());
+        schema.ident_map.insert(i.to_string(), e);
+    }
+
+    fn add_attribute(schema: &mut Schema, e: Entid, a: Attribute) {
+        schema.schema_map.insert(e, a);
+    }
+
+    #[test]
+    fn test_unknown_ident() {
+        let mut cc = ConjoiningClauses::new();
+        let schema = Schema::new();
+
+        cc.apply_pattern(&schema, &Pattern {
+            source: None,
+            entity: PatternNonValuePlace::Variable(Variable(PlainSymbol::new("?x"))),
+            attribute: PatternNonValuePlace::Ident(NamespacedKeyword::new("foo", "bar")),
+            value: PatternValuePlace::Constant(NonIntegerConstant::Boolean(true)),
+            tx: PatternNonValuePlace::Placeholder,
+        });
+
+        assert!(cc.is_known_empty);
+    }
+
+    #[test]
+    fn test_unknown_attribute() {
+        let mut cc = ConjoiningClauses::new();
+        let mut schema = Schema::new();
+
+        associate_ident(&mut schema, ":foo/bar", 99);
+
+        cc.apply_pattern(&schema, &Pattern {
+            source: None,
+            entity: PatternNonValuePlace::Variable(Variable(PlainSymbol::new("?x"))),
+            attribute: PatternNonValuePlace::Ident(NamespacedKeyword::new("foo", "bar")),
+            value: PatternValuePlace::Constant(NonIntegerConstant::Boolean(true)),
+            tx: PatternNonValuePlace::Placeholder,
+        });
+
+        assert!(cc.is_known_empty);
+    }
+
+    #[test]
+    fn test_apply_simple_pattern() {
+        let mut cc = ConjoiningClauses::new();
+        let mut schema = Schema::new();
+
+        associate_ident(&mut schema, ":foo/bar", 99);
+        add_attribute(&mut schema, 99, Attribute {
+            value_type: ValueType::Boolean,
+            ..Default::default()
```

Comment:
Huh, interesting approach.  Cheap builder pattern.  I put this in the TIL of the notes.